### PR TITLE
Improvements on core.authorization and permission backend

### PR DESCRIPTION
**Pull request by leplatrem**
* [x] Refactor and simplify permissions methods to get rid of the callback `get_bound_permissions`
* [x] Rework `get_accessible_objects()` to obtain every objects in a single call (c.f. GET /permissions endpoint)
* [x] Add `get_objects_permissions()` to retrieve the permissions set for a list of objects ids
* [x] Update changelog
* [x] Make tests pass

### CODE REVIEW

File: kinto/core/authorization.py [hunk 1]
```diff
@@ -115,6 +114,8 @@ def __init__(self, request):
         self.get_prefixed_userid = functools.partial(prefixed_userid, request)
 
         self._check_permission = request.registry.permission.check_permission
+        permission = request.registry.permission
```
Comment (leplatrem) [inline kinto/core/authorization.py]: nit: move above

File: kinto/core/authorization.py [hunk 2]
```diff
@@ -179,9 +183,17 @@ def fetch_shared_records(self, perm, principals, get_bound_permissions):
...
         """
-        ids = self.get_shared_ids(permission=perm,
-                                  principals=principals,
-                                  get_bound_permissions=get_bound_permissions)
+        bound_permissions = []
+        if get_bound_permissions is not None:
+            bound_permissions = [(o, p) for (o, p) in
+                                 get_bound_permissions(self._object_id_match,
+                                                       perm)
+                                 if o.endswith(self._object_id_match)]
```
Comment (leplatrem) [inline kinto/core/authorization.py]: Although this was move from every permission backend implementation to here, I believe it is not properly tested / specified

**Comment by leplatrem** [comment]
@Natim feedback?

**Comment by Natim** [comment]
I've read the code I understand that rather than using the get_bound_permissions function you prepare it and then you pass its result.

I think I would need some more explanation but it seems a good refactor to me.

**Comment by leplatrem** [comment]
> using the get_bound_permissions function you prepare it and then you pass its result.

yes, that is it! actually, each backend implementation was resolving the result of this callback the same way! so it makes much more sense to call it before leveraging the backends. It makes the code simpler to read IMO.

Plus, it gives us a consistent way of fetching the permissions for several objects, necessary for #653 and #694 :) Win-Win :)

**Comment by Natim** [comment]
r+

**Pull request merged**